AsyncDisplayKit/Details/ASDataController.mm

```diff
@@ -607,41 +785,46 @@ - (void)reloadRowsAtIndexPaths:(NSArray *)indexPaths withAnimationOptions:(ASDat
   }];
 }
 
-- (void)relayoutAllRows
+- (void)relayoutAllNodes
 {
   [self performEditCommandWithBlock:^{
     ASDisplayNodeAssertMainThread();
     LOG(@"Edit Command - relayoutRows");
     [_editingTransactionQueue waitUntilAllOperationsAreFinished];
-    
-    void (^relayoutNodesBlock)(NSMutableArray *) = ^void(NSMutableArray *nodes) {
-      if (!nodes.count) {
-        return;
-      }
-      
-      [self accessDataSourceWithBlock:^{
-        [nodes enumerateObjectsUsingBlock:^(NSMutableArray *section, NSUInteger sectionIndex, BOOL *stop) {
-          [section enumerateObjectsUsingBlock:^(ASCellNode *node, NSUInteger rowIndex, BOOL *stop) {
-            NSIndexPath *indexPath = [NSIndexPath indexPathForRow:rowIndex inSection:sectionIndex];
-            ASSizeRange constrainedSize = [_dataSource dataController:self constrainedSizeForNodeAtIndexPath:indexPath];
-            [node measureWithSizeRange:constrainedSize];
-            node.frame = CGRectMake(0.0f, 0.0f, node.calculatedSize.width, node.calculatedSize.height);
-          }];
-        }];
-      }];
-    };
 
     // Can't relayout right away because _completedNodes may not be up-to-date,
     // i.e there might be some nodes that were measured using the old constrained size but haven't been added to _completedNodes
     // (see _layoutNodes:atIndexPaths:withAnimationOptions:).
     [_editingTransactionQueue addOperationWithBlock:^{
       ASDisplayNodePerformBlockOnMainThread(^{
-        relayoutNodesBlock(_completedNodes);
+        for (NSString *kind in [_completedNodes keyEnumerator]) {
```

Comment:
Cool, thanks!

> On Oct 18, 2015, at 11:20 PM, appleguy <notifications@github.com> wrote:
> 
> In AsyncDisplayKit/Details/ASDataController.mm:
> 
> >  
> >      // Can't relayout right away because _completedNodes may not be up-to-date,
> >      // i.e there might be some nodes that were measured using the old constrained size but haven't been added to _completedNodes
> >      // (see _layoutNodes:atIndexPaths:withAnimationOptions:).
> >      [_editingTransactionQueue addOperationWithBlock:^{
> >        ASDisplayNodePerformBlockOnMainThread(^{
> > -        relayoutNodesBlock(_completedNodes);
> > +        for (NSString *kind in [_completedNodes keyEnumerator]) {
> @Adlai-Holler you could be right! As it turns out, there are only a very small number of keys in this dictionary anyway, but the thought I had is that NSFastEnumeration is designed to work directly with NSArray & other collections preferentially over enumerator objects.
> 
> Let's leave this code as-is, particularly because either option should not make a measurable difference. Very plausible that for a large dictionary, the enumerator could save memory.
> 
> —
> Reply to this email directly or view it on GitHub.
>